scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -0,0 +1,219 @@
+
+                                 [:expected/killed :pod/succeeded] (do
+                                                                     (pod-has-just-completed kcc existing-state-dict)
+                                                                     (remove-finalization-if-set-and-delete api-client expected-state-dict pod))
+                                 [:expected/killed :pod/failed]
+                                 (do
+                                   (pod-has-just-completed kcc existing-state-dict)
+                                   (remove-finalization-if-set-and-delete api-client expected-state-dict pod))
+                                 [:expected/killed :missing] (pod-was-killed kcc pod-name)
+                                 ; TODO: Implement :pod/unknown cases.
+                                 ; TODO: Implement :pod/pending cases.
+                                 ; TODO: Implement :pod/need-to-fail cases.
+                                 [:missing :pod/running] (kill-task api-client expected-state-dict pod)
+                                 [:missing :pod/pending] (kill-task api-client expected-state-dict pod)
+                                 [:missing :pod/need-to-fail] (kill-task api-client expected-state-dict pod)
+                                 [:missing :pod/succeeded] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 [:missing :pod/failed] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 [:missing :missing] nil ; this can come up due to the recur at the end
+                                 (do
+                                   (log/error "Unexpected state: "
+                                              (vector (or expected-state :missing) (or (:state synthesized-state) :missing))
+                                              "for pod" pod-name)
+                                   expected-state-dict))]
+      (when-not (expected-state-equivalent? expected-state-dict new-expected-state-dict)
+        (update-or-delete! expected-state-map pod-name new-expected-state-dict)
+        (log/info "Processing: WANT TO RECUR" new-expected-state-dict " ---- " existing-state-dict)
+        (recur new-expected-state-dict existing-state-dict)
+        ; TODO: Recur. We hay have changed the expected state, so we should reprocess it.
+        ))))
+
+(defn pod-update
+  "Update the existing state for a pod. Include some business logic to e.g., not change a state to the same value more than once.
+  Invoked by callbacks from kubernetes."
+  [{:keys [existing-state-map] :as kcc} ^V1Pod new-pod]
+  (let [pod-name (api/V1Pod->name new-pod)]
+    (locking (calculate-lock pod-name)
+      (let [new-state {:pod new-pod :synthesized-state (api/pod->synthesized-pod-state new-pod)}
+            old-state (get @existing-state-map pod-name)]
+        (swap! existing-state-map assoc pod-name new-state)
```

Comment:
Comment here: 'We always store the updated state, but only reprocess it if it is genuinely different.'

Reply:
👍